docs/pjrt.md

```diff
@@ -0,0 +1,142 @@
+default configuration, PjRt will create one process per TPU chip, or 4 processes
+per TPU host. See the [Cloud TPU documentation](https://cloud.google.com/tpu/docs/system-architecture-tpu-vm)
+for more information about TPU architecture.
+
+- Performance gains are possible for workloads constrained by data transfer
+  speeds.
+- Under XRT, the server process is the only process that interacts with the TPU
+  devices, and client processes don't have direct access to the TPU devices.
+  When profiling a single-host TPU (e.g. v3-8 or v4-8), you would normally see 8
+  device traces (one for each TPU core). With PjRt, each process has one chip,
+  and a profile from that process will show only 2 TPU cores.
+- For the same reason, profiling does not work on TPU Pods with XRT, because the
+  server process runs independently from the user's model code. PjRt does not
+  have that constraint, so it is possible to profile 2 TPU cores per process in
+  a TPU Pod.
+- PjRt only supports the TPU VM architecture and we have no plans to support the
+  TPU Node architecture with PjRt.
+- Runtime configuration is significantly simpler with PjRt. `xla_dist` is not
+  required to run TPU Pod workloads. Instead, copy your code to each TPU host
+  ([`gcloud compute tpus tpu-vm scp`](https://cloud.google.com/sdk/gcloud/reference/alpha/compute/tpus/tpu-vm/scp)) and run the code on each host in
+  parallel (e.g. [`gcloud compute tpus tpu-vm ssh --workers=all
+  --command="PJRT_DEVICE=TPU python run.py"`](https://cloud.google.com/sdk/gcloud/reference/alpha/compute/tpus/tpu-vm/ssh))
+
+## TPUs v2/v3 vs v4
+
+On TPU v4, one TPU chip is represented to PyTorch as one device, while on TPUs
+v2/v3, one TPU chip is represented to PyTorch as _two_ devices. It is not
+possible to access the same TPU chip from multiple processes, so workloads must
+be able to handle two devices per process. The easiest way to handle this is to
+spawn two threads per process on TPU v2/v3, which is done automatically by
+`xmp.spawn` when using PjRt. With multiple threads per process, multiple replicas
+will share global state, causing the following known issues:
+
+- Threads will share the same `torch` random seed used for parameter
+  initialization. If you relied on each process having the same random seed for
+  deterministic parameter initialization, you will have to synchronize module
+  parameters via collective broadcasting instead (e.g.
+  `pjrt.broadcast_master_param(model)`).
+- `torch.distributed` uses a global process group and does not support
+  multi-threading, so the `xla` `torch.distributed` backend will not work with
+  PjRt and TPU v2 and v3 at this time.
+- Because the current implementation of `xm.rendezvous` for PjRt relies on
+  `torch.distributed`, `xm.rendezvous` is not supported with PjRt on TPU v2 and
+  v3.
+
+### Compatible examples
+
+For an overview of the changes required to migrate from TPU v2/v3 to v4, compare
```

Comment:
Is this part still necessary? I think the old imagenet example works on PJRT right? or does it only works on v4?

Reply:
It only works on v4